AD Helpers: Add helper for vector functions (QP-level)

jppelteret opened the pull request:
Blocked by #6980

**TODO**:
- [x] Update documentation (headers)
- [x] Update documentation (source)
- [x] Check validity of discussions wrt. both `ADOL-C` and `Sacado`
- [x] Proofread

jppelteret commented:
@bangerth I think that this is ready for review, whenever you have time.

jppelteret commented:
@bangerth This still needs to be looked at, when you have time. Afterwards I can add all of the unit tests.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -3280,6 +3280,379 @@ namespace Differentiation
+       * This function is a specialization of the above for rank-0 tensors
+       * (scalars).
+       */
+      Tensor<0, dim, scalar_type>
```
Isn't this just a scalar?

jppelteret replied:
I've made a note in #7849 to do this. I wanted everything to remain consistent (because the extractor helper classes declare  `Tensor<0, dim, scalar_type>` to be the jacobian type. I would want to change this consistently throughout these classes.

bangerth replied:
As mentioned in #7849, I don't feel strongly about it. You make the call.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -3280,6 +3280,379 @@ namespace Differentiation
+       * @f]
+       *
+       * This function is a specialization of the above for rank-4 symmetric
+       * tensors.
```
This corresponds to subsetting a whole set of rows and columns, and then reshaping the matrix so obtained into a `SymmetricTensor<4,dim>`. I like this operation. But can you say something along the lines of my first sentence? I think that would make it clearer.

jppelteret replied:
I've added this to the unspecialised version of the function. I think that its appropriate to put all of the details there (since its the most rigorously documented version of the function).

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -1641,6 +1641,248 @@ namespace Differentiation
+          // operation with respect to the "column" variables,
+          // we only need to consider them for symmetry conditions.
+          if (this->is_symmetric_independent_variable(j) == true)
+            jacobian[i][j] *= 0.5;
```
Uh :-( That's the source of nightmares. Have you tested this on a simple example, preferably with a testcase?

You can make this code faster if you pull the loop over `i` all the way inside.

jppelteret replied:
Yes, I developed this and the counterparts in `ADHelperScalarFunction` using tests. These will all be added in the next PR. I've also been keeping my own repo in sync will all of these changes in order to make sure that none of them break the tests.

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -1641,6 +1641,248 @@ namespace Differentiation
+      // NOTE: It is necessary to make special provision for the case when the
+      // HessianType
+      //       is scalar. Unfortunately Tensor<0,dim> does not provide the
+      //       function unrolled_to_component_indices!
```
funny indentation. On purpose?

bangerth replied:
Also, there is no reason not to add that function to `Tensor<0>` if that would make your life easier!

jppelteret replied:
I've made a note in #7665 to do this.

jppelteret replied:
> funny indentation. On purpose?

Funny question :-D No, a certain tool (which shall remain anonymous) did something without me noticing...

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -1641,6 +1641,248 @@ namespace Differentiation
+      // this class. NOTE: We require a specialisation for rank-4 symmetric
+      // tensors because they
+      //       do not define their rank, and setting data using TableIndices is
+      //       somewhat specialised as well.
```
same here

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -1641,6 +1641,248 @@ namespace Differentiation
+        const FEValuesExtractors::SymmetricTensor<2> &extractor_col) const
+    {
+      // NOTE: The order of components must be consistently defined throughout
+      // this class. NOTE: We require a specialisation for rank-4 symmetric
```
I think you want a line break before the second NOTE

bangerth commented:
By the way, are all of these classes prefixed with `ADHelper`? Are there non-helper classes anywhere? Because, if so, maybe that could be put into the name of a namespace, or dropped altogether.

A `FEValues` is really also just a `FEHelper`. I'm not sure where I'm going with this sentence, it just popped in my mind when trying to associate our usual naming practices :-)

jppelteret commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -3006,7 +3006,7 @@ namespace Differentiation
      *     // of the extractors that we've created.
      *     const SymmetricTensor<2,dim,ADNumberType> C_AD =
-     *       ad_helper.get_sensitive_variables(C_dofs); const
+     *       ad_helper.get_sensitive_variables(C_dofs);
```
Yeah, not sure how that snuck in...

jppelteret commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -3280,6 +3280,379 @@ namespace Differentiation
+     *   // computed.
+     *   const Tensor<3,dim,double> PP =
+     *     -ad_helper.extract_jacobian_component(jacobian,S_dofs,H_dofs)
+     * @endcode
```
Thanks :-)

jppelteret commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -3280,6 +3280,379 @@ namespace Differentiation
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType = double>
+    class ADHelperVectorFunction
+      : public ADHelperPointLevelFunctionsBase<dim,
```
Maybe it is a little too descriptive :-D

jppelteret commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -3280,6 +3280,379 @@ namespace Differentiation
+      typename internal::VectorFieldValue<dim, scalar_type, ExtractorType_Row>::
+        type
+        extract_value_component(const Vector<scalar_type> &values,
+                                const ExtractorType_Row &  extractor_row) const;
```
True... I'll mark them as such.

bangerth replied:
Do you think they might be of more general interest than just here? If they really are static, I could see these kinds of functions be useful in `namespace FEValuesExtractors` where the compiler will find them due to ADL.

Not necessarily as part of this patch, but maybe a follow-up if you think this is useful?

jppelteret commented:
> By the way, are all of these classes prefixed with ADHelper? Are there non-helper classes anywhere? Because, if so, maybe that could be put into the name of a namespace, or dropped altogether.

See #7848

jppelteret commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -3280,6 +3280,379 @@ namespace Differentiation
+       *
+       * @param[out] jacobian A FullMatrix with the gradient of each component
+       * of the vector field evaluated at the point defined by the independent
+       * variable values.
```
I've added a note that is consistent with the `compute_linearization` etc. functions of the cell-level helpers. I hope this is satisfactory.

bangerth replied:
Yes.

jppelteret commented:
Thanks a lot for reviewing this @bangerth . If you approve of the changes then I'll squash as many of the commits as I can before we merge this PR.

bangerth commented:
Yes, there was nothing much I had in my comments. So please squash and merge yourself once the tests are through! Thanks as usual for the massive amount of work. Also special thanks for the extraordinary job documenting this!

jppelteret commented:
> Thanks as usual for the massive amount of work. 

Thanks a lot to you as well for the thorough feedback you've given me!

> Also special thanks for the extraordinary job documenting this!

Its a pleasure. I can only hope that I'm not confused when trudging though this in a few days/weeks/months/years time :-D

jppelteret commented:
> So please squash and merge yourself once the tests are through!

I've squashed the changes and added a changelog entry. I still need one click on the approve button before its possible to merge ;-)

jppelteret commented:
/rebuild

bangerth commented:
Yay -- making progress with the AD stuff!

jppelteret commented:
> Yay -- making progress with the AD stuff!

Pretty much done with the AD stuff... 🎉🎉🎉
